ncm-spma/src/main/perl/spma/apt.pm

```diff
@@ -0,0 +1,386 @@
+
+A single SPMA run consists of the following steps:
+
+=over
+
+=item Setup source directory if required
+
+=item Remove sources that are not found in the profile
+
+=item Update source cache from upstream sources
+
+=item Upgrade already installed packages
+
+=item Install packages specified in the profile that are not installed
+
+=item Mark any packages installed but not in the profile as automatically installed
+
+=item Ask apt to remove all automatically installed packages that are not satisfying dependencies of other packages
+
+=back
+
+=head1 RESOURCES
+
+Only a very minimal schema is implemented.
+
+Sources listed under C</software/repositories> will be configured,
+URLs should be followed by the suite and sections required e.g. C<http://example.org/debian unstable main>
+
+Packages listed under C</software/packages> will be installed, version and architecture locking (including multiarch) is fully implemented.
+
+=cut
+
+use strict;
+use warnings;
+use parent qw(NCM::Component CAF::Path);
+use CAF::Process;
+use CAF::FileWriter;
+use CAF::FileEditor;
+use EDG::WP4::CCM::Path qw(escape unescape);
+use EDG::WP4::CCM::TextRender;
+use NCM::Component::spma::yum;
+use CAF::Object qw(SUCCESS);
+use Set::Scalar;
+use File::Path qw(mkpath);
+use Text::Glob qw(match_glob);
+use Readonly;
+
+Readonly my $DIR_SOURCES => "/etc/apt/sources.list.d";
+Readonly my $DIR_PREFERENCES => "/etc/apt/preferences.d";
+Readonly my $TEMPLATE_SOURCES => "apt/source.tt";
+Readonly my $TEMPLATE_PREFERENCES => "apt/preferences.tt";
+Readonly my $TEMPLATE_CONFIG => "apt/config.tt";
+Readonly my $FILE_CONFIG => "/etc/apt/apt.conf.d/98quattor.conf";
+Readonly my $TREE_SOURCES => "/software/repositories";
+Readonly my $TREE_PKGS => "/software/packages";
+Readonly my $BIN_APT_GET => "/usr/bin/apt-get";
+Readonly my $BIN_APT_MARK => "/usr/bin/apt-mark";
+Readonly my $BIN_DPKG_QUERY => "/usr/bin/dpkg-query";
+Readonly my $CMD_APT_UPDATE => [$BIN_APT_GET, qw(-qq update)];
+Readonly my $CMD_APT_UPGRADE => [$BIN_APT_GET, qw(-qq dist-upgrade)];
+Readonly my $CMD_APT_INSTALL => [$BIN_APT_GET, qw(-qq install)];
+Readonly my $CMD_APT_AUTOREMOVE => [$BIN_APT_GET, qw(-qq autoremove)];
+Readonly my $CMD_APT_MARK => [$BIN_APT_MARK, qw(-qq)];
+Readonly my $CMD_DPKG_QUERY => [$BIN_DPKG_QUERY, qw(-W -f=${db:Status-Abbrev};${Package}\n)];
+
+our $NoActionSupported = 1;
+
+
+# If user specified sources (userrepos) are not allowed, removes any
+# sources present in the system that are not listed in $allowed_sources.
+sub cleanup_old_sources
+{
+    my ($self, $sources_dir, $allowed_sources) = @_;
+    $self->debug(5, 'Entered cleanup_old_sources()');
+
+    my $current = Set::Scalar->new(glob("$sources_dir/*.list"));
+    my $allowed = Set::Scalar->new(map($_->{name}, @$allowed_sources));
+    my $to_remove = $current-$allowed;
+
+    if ($self->directory_exists($sources_dir)) {
```

Comment:
what test is failing? you also need latest build-tools and everything has to be mocked, but should all work